### ISSUE TITLE
Contract unable to verify signed message on Rinkeby (invalid V value)

### ISSUE THREAD

**Issue by vilius** [q9f/eth.rb#45]
Observation:

Typed data signature using eth.rb vs ethers.js differ in last bytes (both set to Chain ID 4). eg:
```
// ethers.js
0x3e2a2ced47f08d207d3fdf7804e20ef56cff751e3f0ce18e49bee84c9d914e06216e9623aa140643bbcac707498b63ec2ad26654d21ce0d83cab781d0cfb71021b

// eth.rb
0x3e2a2ced47f08d207d3fdf7804e20ef56cff751e3f0ce18e49bee84c9d914e06216e9623aa140643bbcac707498b63ec2ad26654d21ce0d83cab781d0cfb71022b
```

When sent to a contract on Rinkeby `ethers.js` signature can be verified, but `eth.rb` fails with message `invalid 'v' value`.
Interestingly `ethers.js` can verify both messages and return correct address.

In the code I've noticed following implementation:
```ruby
# Converts a recovery ID into the expected `v` on a given chain.
#
# @param recovery_id [Integer] signature recovery id.
# @param chain_id [Integer] the chain id the signature was generated on.
# @return [Integer] the signature's `v` value.
def to_v(recovery_id, chain_id = ETHEREUM)
  v = 2 * chain_id + 35 + recovery_id
end
```

EIP-155 says:

> If block.number >= FORK_BLKNUM and CHAIN_ID is available, then when computing the hash of a transaction for the purposes of signing, instead of hashing only six rlp encoded elements (nonce, gasprice, startgas, to, value, data), you SHOULD hash nine rlp encoded elements (nonce, gasprice, startgas, to, value, data, chainid, 0, 0). If you do, then the v of the signature MUST be set to {0,1} + CHAIN_ID * 2 + 35 where {0,1} is the parity of the y value of the curve point for which r is the x-value in the secp256k1 signing process. If you choose to only hash 6 values, then v continues to be set to {0,1} + 27 as previously.

Following change fixed the issue:
```diff
- v = 2 * chain_id + 35 + recovery_id
+ v = 27 + recovery_id
```

However I'm 100% uncertain if that's a fix or I'm simply missing something else?

Thanks for your work on this!

**Comment by q9f** [comment]
Thank you for your report.

While I implemented this, I was aware that I did not add any way to support pre-155 legacy values for `v`.

The fix you proposed would break `v` values for any signature that respects EIP-155.

I'm surprised that you are using Chain ID of 4 but still get a `v` value not respecting the ID. Did you talk to the ethers.js developers about this? 

While I acknowledge that legacy support is still missing, I would argue, that a signature with chain ID cannot have a `v` value of 27 or 28.

**Comment by q9f** [comment]
> When sent to a contract on Rinkeby ethers.js signature can be verified, but eth.rb fails with message invalid 'v' value.
Interestingly ethers.js can verify both messages and return correct address.

I have misread this bit. Do you mean the `invalid 'v' value` message comes from the contract? Could you show me the contract and signed data, so that I can write a test case?

**Comment by q9f** [comment]
I made a clean fix in #47 that allows to give you control over the chain ID. By not specifying it, you automatically get a legacy value for `v` (27, 28).
